### PR TITLE
add initial support for ovs-appctl ofproto/trace

### PR DESCRIPTION
**Pull request by andrewsykim**
Adds support to parse data path actions based on ofproto/trace output.

### CODE REVIEW

File: ovs/app.go [hunk 1]
```diff
@@ -0,0 +1,58 @@
+// Copyright 2017 DigitalOcean.
```
[the remark this answers is not in the document]
Comment (andrewsykim) [reply ovs/app.go]: There's a static check somewhere in CI that checks for 2017 specifically so I didn't bother trying to update the entire code base here 😆

[the remark this answers is not in the document]
Comment (ecbaldwin) [reply ovs/app.go]: That sounds like a silly self-inflicted problem that is easily solved in another quick PR. :)

File: ovs/proto_trace.go [hunk 2]
```diff
@@ -0,0 +1,192 @@
...
+func (d *dataPathAction) UnmarshalText(b []byte) error {
+	action := string(b)
+	if _, err := strconv.Atoi(action); err == nil {
+		// data path action is to an output port
+		d.action = action
+		return nil
+	}
+
+	if strings.HasPrefix(action, pushvlan) {
+		d.action = pushvlan
```
Comment (ecbaldwin) [inline ovs/proto_trace.go]: The value of the vlan getting pushed with this action is lost here. I'm wondering if a single dataPathAction struct is expressive enough to show all of the possibilities. I'm wondering if there should be a different struct for each kind of data path action with fields specific to that type. For example, there would be a PushVlanAction struct with the actual vlan (and maybe pcp, etc) as members of that struct. There would also be an OutputPortAction which would have the port number as a member of the struct. Thoughts?

Another complicated action that I've seen is: `set(eth(src=fe:00:00:00:01:01,dst=96:61:6a:c7:d8:ec))` I'm having trouble figuring out the best way to express that with this. I would do it something like this:

type SetEthernetAction struct {
    Src string
    Dst string
}

However, this breaks the UnmarshalText pattern, so we'd have to do it differently.

File: ovs/app.go [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+
+// AppService runs commands that are available from ovs-appctl
+type AppService struct {
+	c *Client
+}
+
+// ProtoTrace runs ovs-appctl ofproto/trace on the given bridge and match flow
+// and returns a *ProtoTrace. Also returns err if there is any error parsing the
+// output from ovs-appctl ofproto/trace.
+func (a *AppService) ProtoTrace(bridge string, matches []Match) (*ProtoTrace, error) {
```
[the remark this answers is not in the document]
Comment (andrewsykim) [reply ovs/app.go]: noted, I don't have a better alternative to this at the moment but I'll circle back here :)

File: ovs/proto_trace.go [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+
+		df.Matches = append(df.Matches, m)
+	}
+
+	return nil
+}
+
+// ProtoTrace is a type representing output from ovs-app-ctl ofproto/trace
+type ProtoTrace struct {
+	InputFlow       *DataPathFlows
```
Comment (andrewsykim) [inline ovs/proto_trace.go]: fwiw, InputFlow and FinalFlow may not even be needed.. but I'm gonna leave this for now.

File: ovs/matchparser.go [hunk 5]
```diff
@@ -248,6 +277,25 @@ func parseDataLinkVLAN(value string) (Match, error) {
 	return DataLinkVLAN(int(vlan)), nil
 }
 
+// parseDataLinkVLANPCP parses a DataLinkVLAN Match from value.
```
Comment (icosahedral) [inline ovs/matchparser.go]: Believe you want "parses a DataLinkVLANPCP Match from value."

File: ovs/proto_trace.go [hunk 6]
```diff
@@ -0,0 +1,173 @@
...
+	"regexp"
+	"strings"
+)
+
+var (
+	// ErrInvalidProtoTrace is returned when the output from
+	// ovs-appctl ofproto/trace is in an unexpected format
+	ErrInvalidProtoTrace = errors.New("invalid ofproto/trace output")
+
+	datapathActionsRegexp = regexp.MustCompile(`Datapath actions: (.*)`)
```
Comment (icosahedral) [inline ovs/proto_trace.go]: I'm _pretty_ sure these specific lines have been consistent, but we might want to sanity check versions before attempting to parse.  I know at least the individual trace of each table inside of the pipeline has changed, for instance.

File: ovs/proto_trace.go [hunk 7]
```diff
@@ -0,0 +1,173 @@
...
+	pushvlan = "pushvlan"
+	drop     = "drop"
+)
+
+// DataPathActions is a text unmarshaler for data path actions in ofproto/trace output
+type DataPathActions interface {
+	encoding.TextUnmarshaler
+}
+
+// NewDataPathActions returns an implmenetation of DataPathActions
```
Comment (icosahedral) [inline ovs/proto_trace.go]: nit: implementation

File: ovs/proto_trace.go (hunk 4 above)
Comment (icosahedral) [reply ovs/proto_trace.go]: Yeah - might be good to at least warn if they differ significantly on test cases, just so we know we're explicit with what we're exercising.

**Comment by andrewsykim** [comment]
`staticcheck` is failing and flagging things I haven't changed in this PR 
```
$ staticcheck ./...
ovs/flow.go:33:35: error strings should not be capitalized (ST1005)
ovs/flow.go:37:2: var errPriorityNotFirst is unused (U1000)
ovs/match.go:30:2: const sourceHardwareAddr is unused (U1000)
ovs/match.go:31:2: const targetHardwareAddr is unused (U1000)
ovs/match.go:32:2: const sourceProtocolAddr is unused (U1000)
ovs/match.go:33:2: const targetProtocolAddr is unused (U1000)
ovs/openflow.go:312:2: var dumpAggregatePrefix is unused (U1000)
ovs/portrange.go:44:5: no value of type uint16 is less than 0 (SA4003)
ovs/portrange.go:44:21: no value of type uint16 is less than 0 (SA4003)
ovs/proto_trace.go:34:2: var pushVLANPattern is unused (U1000)
ovs/proto_trace.go:38:2: const popvlan is unused (U1000)
ovs/proto_trace.go:39:2: const pushvlan is unused (U1000)
ovs/proto_trace.go:40:2: const drop is unused (U1000)
ovsdb/client_integration_test.go:58:30: should use make(chan struct{}) instead (S1019)
ovsnl/internal/ovsh/const.go:122:6: type ovsDatapathCmd is unused (U1000)
ovsnl/internal/ovsh/const.go:134:6: type ovsDatapathAttr is unused (U1000)
ovsnl/internal/ovsh/const.go:149:6: type ovsPacketCmd is unused (U1000)
ovsnl/internal/ovsh/const.go:160:6: type ovsPacketAttr is unused (U1000)
ovsnl/internal/ovsh/const.go:179:6: type ovsVportCmd is unused (U1000)
ovsnl/internal/ovsh/const.go:191:6: type ovsVportType is unused (U1000)
ovsnl/internal/ovsh/const.go:205:6: type ovsVportAttr is unused (U1000)
ovsnl/internal/ovsh/const.go:243:6: type ovsFlowCmd is unused (U1000)
ovsnl/internal/ovsh/const.go:255:6: type ovsKeyAttr is unused (U1000)
ovsnl/internal/ovsh/const.go:292:6: type ovsTunnelKeyAttr is unused (U1000)
ovsnl/internal/ovsh/const.go:316:6: type ovsFragType is unused (U1000)
ovsnl/internal/ovsh/const.go:327:6: type ovsNshKeyAttr is unused (U1000)
ovsnl/internal/ovsh/const.go:339:6: type ovsFlowAttr is unused (U1000)
ovsnl/internal/ovsh/const.go:359:6: type ovsSampleAttr is unused (U1000)
ovsnl/internal/ovsh/const.go:370:6: type ovsUserspaceAttr is unused (U1000)
ovsnl/internal/ovsh/const.go:383:6: type ovsHashAlg is unused (U1000)
ovsnl/internal/ovsh/const.go:391:6: type ovsCtAttr is unused (U1000)
ovsnl/internal/ovsh/const.go:408:6: type ovsNatAttr is unused (U1000)
ovsnl/internal/ovsh/const.go:426:6: type ovsActionAttr is unused (U1000)
ovsnl/internal/ovsh/const.go:454:6: type ovsMeterCmd is unused (U1000)
ovsnl/internal/ovsh/const.go:466:6: type ovsMeterAttr is unused (U1000)
ovsnl/internal/ovsh/const.go:484:6: type ovsBandAttr is unused (U1000)
ovsnl/internal/ovsh/const.go:497:6: type ovsMeterBandType is unused (U1000)
The command "staticcheck ./..." exited with 1.
```

**Comment by andrewsykim** [comment]
There was a new version of staticcheck recently https://github.com/dominikh/go-tools/releases/tag/2019.1 which I think fails for our codebase.

**Pull request merged**